Hazelcast event table extension

grainier opened the pull request:
Hazelcast event table extension will allow Siddhi to hold large number of events in-memory in a distributed manner.

lasanthafdo commented on modules/siddhi-extensions/event-table/pom.xml:
```diff
@@ -76,23 +95,40 @@
         <plugins>
             <plugin>
                 <groupId>org.apache.felix</groupId>
+                <artifactId>maven-scr-plugin</artifactId>
```
@grainier 
Have we defined the scr-plugin anywhere in the parent? If not, you need to provide all the configuration details here properly. i.e. the version, goals need to be specified. You can specify them in pluginManagement section and then just declare the group id and artifact id in the plugin section.

lasanthafdo commented on modules/siddhi-extensions/event-table/pom.xml:
```diff
@@ -76,23 +95,40 @@
                             org.apache.hadoop.util.bloom,
                             org.apache.hadoop.util.bloom.*,
+                            org.osgi.framework,
                             *;resolution:=optional
```
Can you try to remove this resolution:=optional directive and include the needed dependencies with versions directly? It is recommended to provide the direct imports with versions whenever possible.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/hazelcast/HazelcastOperatorParser.java:
```diff
@@ -0,0 +1,227 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
```
Grainier,
Can you provide class level comments for this?

grainier replied:
Fixed in #114

lasanthafdo commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/hazelcast/HazelcastOperatorParser.java:
```diff
@@ -0,0 +1,227 @@
+            matchingStreamIndex = 0;
+            streamEventSize = 2;
+        } else {
+
```
Better remove any unnecessary new lines.

grainier commented:
@lasanthafdo 
1. scr-plugin removed since it's no longer required (because now it uses osgi annotations instead of javadoc scr tags). 
2. `resolution:=optional` directive was removed and added the needed dependencies with versions.

@dilini-muthumala 
1. Version of Siddhi, Hazelcast, OSGi were added to the parent pom as properties.

dilini-muthumala commented:
@grainier As discussed offline, let's move hazelcast dependency (and others with version element) under dependency management section in the super pom (which is Siddhi pom). After doing that, we don't need to specify versions in siddhi-extension-event-table pom.

grainier commented:
@lasanthafdo 
1. Added class level comment to HazelcastOperatorParser
2. Removed unnecessary new lines in HazelcastOperatorParser.

@dilini-muthumala 
1. Rearranged dependencies as mentioned above

dilini-muthumala commented on modules/siddhi-extensions/event-table/src/main/resources/hazelcast-table-config.xml:
```diff
@@ -0,0 +1,16 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
```
@grainier Let's update this license header as well to the correct one.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/IndexedTableTestCase.java:
```diff
@@ -0,0 +1,423 @@
+                                Assert.assertArrayEquals(new Object[]{"IBM", 300l}, event.getData());
+                                break;
+                            default:
+                                Assert.assertSame(1, inEventCount.get());
```
@grainier 
Since you are already handling the case of inEventCount.get() == 1 above, I think if the code hits the default section, it will already be something other than 1. So Assert.assertSame will always result in failure. In such a case, you can use the Assert.fail() instead of doing an additional check.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/IndexedTableTestCase.java:
```diff
@@ -0,0 +1,423 @@
+                        inEventCount.incrementAndGet();
+                        switch (inEventCount.get()) {
+                            case 1:
+                                Assert.assertArrayEquals(new Object[]{"IBM", 300l}, event.getData());
```
@grainier 
This assertion is coming from the callback implementation. The assertion failure will be triggered from the thread that calls the receive() method and not from the thread running the test case. You need to move the assertion out of the callback so that the test case will fail in such a scenario. Please fix this if there are any other instances such as this as well.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/InsertIntoTableTestCase.java:
```diff
@@ -0,0 +1,458 @@
+                                Assert.assertArrayEquals(new Object[]{"WSO2"}, event.getData());
+                                break;
+                            default:
+                                Assert.assertSame(1, inEventCount.get());
```
Can you refactor these assertions as well so that the assertion failure happens within the test case thread?

grainier commented:
@dilini-muthumala 
1. Removed hazelcast-table-config.xml since it's no longer required. 

@lasanthafdo 
1. In test cases, moved assertions out of the callback implementation.

mohanvive commented on modules/siddhi-core/src/main/java/org/wso2/siddhi/core/table/InMemoryEventTable.java:
```diff
@@ -41,17 +41,14 @@
-import java.util.Map;
-import java.util.TreeMap;
+import java.util.*;
 
```
Always it is welcomed to have direct imports. Rather having global import.

grainier replied:
@mohanvive Replaced with single class imports in #109

sacjaya commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/HazelcastEventTable.java:
```diff
@@ -0,0 +1,308 @@
+    private ExecutionPlanContext executionPlanContext;
+    private StreamEventCloner streamEventCloner;
+    private StreamEventPool streamEventPool;
+    private HazelcastInstance hcInstance;
```
This can be converted to a local variable.

grainier replied:
@sacjaya fixed in #109

lasanthafdo commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/HazelcastEventTable.java:
```diff
@@ -0,0 +1,308 @@
+            if (HazelcastEventTableServiceValueHolder.getHazelcastInstance() != null) {
+                // take instance from osgi
+                hazelcastInstance = HazelcastEventTableServiceValueHolder.getHazelcastInstance();
+                logger.info("shared hazelcast server instance retrieved : " + hazelcastInstance.getName());
```
@grainier shall we fix the capitalization for the log messages as well?

grainier replied:
@lasanthafdo Fixed in #114.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/DefineTableTestCase.java:
```diff
@@ -0,0 +1,471 @@
+
+
+public class DefineTableTestCase {
+    static final Logger log = Logger.getLogger(DefineTableTestCase.class);
```
Let's put access modifiers properly as a general best practice.

grainier replied:
@lasanthafdo Fixed in #114.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/DefineTableTestCase.java:
```diff
@@ -0,0 +1,471 @@
+            hciNames.add(hci.getName());
+        }
+        Assert.assertTrue(hciNames.contains(HazelcastEventTableConstants.HAZELCAST_INSTANCE_PREFIX + executionPlanRuntime.getName()));
+        executionPlanRuntime.shutdown();
```
@grainier Better call the ExecutionPlanRuntime.shutdown() method within a finally block as a best practice so that assertion failures do not cause leaking of resources.

grainier replied:
@lasanthafdo Fixed in #114.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/DefineTableTestCase.java:
```diff
@@ -0,0 +1,471 @@
+        executionPlanRuntime.start();
+        stockStream.send(new Object[]{"WSO2", 55.6f});
+        stockStream.send(new Object[]{"IBM", 75.6f});
+        Thread.sleep(RESULT_WAIT * 4);
```
@grainier Shall we introduce the 'waitForEvents' logic here as well so that the probability of test cases failing due to slow machines is low?

grainier replied:
@lasanthafdo Fixed in #114.

lasanthafdo commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/DefineTableTestCase.java:
```diff
@@ -0,0 +1,471 @@
+        Assert.assertTrue(instanceMap.containsKey(HazelcastEventTableConstants.HAZELCAST_INSTANCE_PREFIX + executionPlanRuntime.getName()));
+        HazelcastInstance instance = instanceMap.get(HazelcastEventTableConstants.HAZELCAST_INSTANCE_PREFIX + executionPlanRuntime.getName());
+        List<StreamEvent> streamEvents = instance.getList(HazelcastEventTableConstants.HAZELCAST_LIST_INSTANCE_PREFIX + executionPlanRuntime.getName() + '_' + tableDefinition.getId());
+        Assert.assertEquals(2, streamEvents.size());
```
Since 'isEventsMatch' utility method is available, can we check whether the list has the correct events as well in addition to checking whether the number of events are correct?

grainier replied:
@lasanthafdo Fixed in #114.

keizer619 commented on modules/siddhi-core/src/test/java/org/wso2/siddhi/core/query/join/JoinTestCase.java:
```diff
@@ -185,61 +185,62 @@ public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
-    }
+    // TODO : TimeWindowProcessor should be fixed in order to run below test case
+//    @Test
+//    public void joinTest4() throws InterruptedException {
```
@grainier 
Why this test method is commented? If it is not needed, could we remove it?

grainier replied:
@keizer619 This Test case is required. However, due to a [bug] (https://wso2.org/jira/browse/CEP-1413) in TimeWindowProcessor, test case temporarily disabled (by commenting out `@Test` annotation in #114.). [New jira] (https://wso2.org/jira/browse/CEP-1452) was created to re-enable the test case, once TimeWindowProcessor get fixed.

sinthuja replied:
You could disable the test case with (enabled=false) rather commenting entirely..

grainier replied:
@sinthuja Fixed in #116 using JUnit `@Ignore` annotation.

tishan89 commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/HazelcastEventTable.java:
```diff
@@ -0,0 +1,308 @@
+    private StreamEventCloner streamEventCloner;
+    private StreamEventPool streamEventPool;
+    private HazelcastInstance hcInstance;
+    private ConcurrentMap<Object, StreamEvent> eventsMap = null;
```
Any specific reason to declare this as ConcurrentMap. If not please declare as Map<K,V>

grainier replied:
@tishan89 reasons;
1. Hazelcast Map (`IMap`) extends the interface `java.util.concurrent.ConcurrentMap`.
2. Since it is the distributed, might require concurrent modification of the Map from several threads without the need to block them.

tishan89 commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/hazelcast/HazelcastOperatorParser.java:
```diff
@@ -0,0 +1,215 @@
+        }
+        return null;
+    }
+}
```
There are several places in class where you exceed line width of 200. Please adhere to WSO2 coding styles

grainier replied:
@tishan89 Fixed in #114.

tishan89 commented on modules/siddhi-extensions/event-table/src/test/java/org/wso2/siddhi/extension/eventtable/hazelcast/DefineTableTestCase.java:
```diff
@@ -0,0 +1,471 @@
+        Thread.sleep(RESULT_WAIT * 4);
+
+        List<StreamEvent> streamEvents = instance_2.getList(HazelcastEventTableConstants.HAZELCAST_LIST_INSTANCE_PREFIX + executionPlanRuntime.getName() + '_' + tableDefinition.getId());
+        Assert.assertEquals(2, streamEvents.size());
```
Exceeding line width

grainier replied:
@tishan89 Fixed in #114.

dilini-muthumala commented on modules/siddhi-core/src/test/java/org/wso2/siddhi/core/query/join/JoinTestCase.java:
```diff
@@ -185,61 +185,62 @@ public void receive(long timeStamp, Event[] inEvents, Event[] removeEvents) {
-        executionPlanRuntime.shutdown();
-
-    }
+    // TODO : TimeWindowProcessor should be fixed in order to run below test case
```
@grainier I think it's better to include the Jira ID for the issue w.r.t the TimeWindowProcessor, in the comment itself; and why not create a Jira to track this, otherwise we might miss to fix this testcase. WDYT?

grainier replied:
@dilini-muthumala In #114, added JIRA and referenced them in comment.
1. JIRA for the TimeWindowProcessor bug (https://wso2.org/jira/browse/CEP-1413).
2. JIRA for re-enabling the test case (https://wso2.org/jira/browse/CEP-1452).

dilini-muthumala commented on modules/siddhi-extensions/event-table/src/main/java/org/wso2/siddhi/extension/eventtable/HazelcastEventTable.java:
```diff
@@ -0,0 +1,308 @@
+        this.tableDefinition = tableDefinition;
+        this.executionPlanContext = executionPlanContext;
+
+        Annotation fromAnnotation = AnnotationHelper.getAnnotation(SiddhiConstants.ANNOTATION_FROM, tableDefinition.getAnnotations());
```
@grainier Possible NPE, i.e. if the user had not used "from" annotation.

grainier replied:
@dilini-muthumala in-order to use hazelcast event table, user has to define it as;
```
@from(eventtable = 'hazelcast')
define table OutputStream (symbol string, price int, volume float);
```
And if the `@from` annotation is not present, it'll function as a normal In-memory event table.

Kishanthan commented on modules/siddhi-extensions/event-table/pom.xml:
```diff
@@ -86,14 +99,29 @@
+                            org.apache.hadoop.util.hash;version="${hadoop.common.version}",
+                            org.apache.log4j;version="${log4j.version}",
+                            org.apache.tomcat.jdbc.pool;version="${jdbc-pool.version}",
+                            javax.sql,
```
import packages should have version range

grainier replied:
@Kishanthan it was already fixed in #81

lalaji commented on modules/siddhi-core/src/main/java/org/wso2/siddhi/core/table/InMemoryEventTable.java:
```diff
@@ -41,17 +41,14 @@
     private final ExecutionPlanContext executionPlanContext;
-    private LinkedList<StreamEvent> list = new LinkedList<StreamEvent>();
-    private TreeMap<Object, StreamEvent> treeMap = new TreeMap<Object, StreamEvent>();
+    private List<StreamEvent> list = new LinkedList<StreamEvent>();
```
It's better to add proper names for variables,rather generic terms as list,treemap

grainier replied:
@lalaji Fixed in #121.
